src/main/java/com/aws/greengrass/componentmanager/plugins/DockerImageDownloader.java

```diff
@@ -0,0 +1,155 @@
+            try {
+                RetryUtils.runWithRetry(retryConfig, () -> {
+                    AuthorizationData authorizationData = ecrAccessor.getCredentials(image.getRegistry().getEndpoint());
+                    // Decoded auth token is of the format <username>:<password>
+                    String[] authTokenParts =
+                            new String(Base64.decodeBase64(authorizationData.authorizationToken())).split(":");
+                    image.registry.setCredentials(new Registry.Credentials(authTokenParts[0], authTokenParts[1],
+                            authorizationData.expiresAt()));
+                    return null;
+                }, "get-ecr-auth-token", logger);
+            } catch (InterruptedException e) {
+                throw e;
+            } catch (Exception e) {
+                throw new PackageDownloadException(getErrorString("Failed to get auth token for docker login"), e);
+            }
+
+            // Login to registry
+            try {
+                RetryUtils.runWithRetry(retryConfig, () -> dockerCliClient.login(image.getRegistry()), "docker-login",
+                        logger);
+            } catch (InterruptedException e) {
+                throw e;
+            } catch (Exception e) {
+                throw new PackageDownloadException(getErrorString("Failed to login to docker registry"), e);
+            }
+        }
+
+        // Docker pull
+        // TODO : Redo credential fetching and login if ECR registry credentials expire by the time pull is attempted
+        try {
+            RetryUtils.runWithRetry(retryConfig, () -> dockerCliClient.pullImage(image), "docker-pull-image", logger);
+        } catch (InterruptedException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new PackageDownloadException(getErrorString("Failed to download docker image"), e);
+        }
+
+        // No file resources available since image artifacts are stored in docker's image store
+        return null;
+    }
+
+    @Override
+    protected long download(long rangeStart, long rangeEnd, MessageDigest messageDigest)
+            throws PackageDownloadException, InterruptedException {
+        // N/A since handling partial download is managed by docker engine
+        return 0;
+    }
+
+    @Override
+    public boolean downloadRequired() {
+        // TODO : Consider executing `docker images` to see if the required image version(tag/digest) already
+        //  exists to save a download attempt
+        return false;
+    }
+
+    @Override
+    public Optional<String> checkDownloadable() {
+        // TODO : Maybe worth checking if device is configured such that it can get TES credentials for ECR.
+        // N/A for images from other registries
+        return Optional.empty();
+    }
+
+    @Override
+    public Long getDownloadSize() throws PackageDownloadException, InterruptedException {
+        // Not supported for docker images
+        return null;
+    }
```

Comment:
Do it mean customers are required to clean up docker images manually?

Reply:
For now, that's the decision. We discussed certain ways to address automatic cleanup, but it'll be a follow up item